File: linked-lists/8.c
/**
 * 8.c
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements a priority queue abstraction
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

#include <stdio.h>
#include <stdlib.h>

typedef struct node
{
    int value;
    int priority;
    struct node *next;
} node;

// conjoins a new node
node* insert (node **head, int value, int priority);

// removes the highest priority node
node* serve (node **head);

// returns the highest priority node
node* peek (node **head);

// computes the node count
int range (node *head);

int main (void)
{

    node *head = NULL;

    insert(&head, 25, 20);
    insert(&head, 50, 30);
    insert(&head, 75, 25);

    printf("%i\n", range(head));
    printf("%i\n", peek(&head) -> value);

    free(serve(&head));

    node *elect = head;

    while (elect != NULL)
    {
        printf("%i\t%i\n", elect -> value, elect -> priority);
        elect = elect -> next;
    }

    while (head != NULL)
    {
        elect = head;
        head = head -> next;
        free(elect);
    }

}

node* insert (node **head, int value, int priority)
{

    node *elect = *head;
    node *inset = malloc(sizeof(node));

    if (inset == NULL)
    {
        return NULL;
    }

    inset -> value = value;
    inset -> priority = priority;
    inset -> next = NULL;

    if (elect == NULL)
    {
        return *head = inset;
    }

    node *previous = NULL;

    while (elect != NULL)
    {

        if (elect -> priority < priority)
        {
            break;
        }

        previous = elect;
        elect = elect -> next;

    }

    if (previous == NULL)
    {
        inset -> next = elect;
        return *head = inset;
    }

    previous -> next = inset;
    inset -> next = elect;

    return *head;

}

node* serve (node **head)
{

    node *elect = *head;

    if (elect == NULL)
    {
        return NULL;
    }

    *head = elect -> next;

    return elect;

}

node* peek (node **head)
{
    return *head;
}

int range (node *head)
{

    int n = 0;

    while (head != NULL)
    {
        head = head -> next;
        n += 1;
    }

    return n;

}

// vim: set expandtab shiftwidth=4:

File: linked-lists/8.js
/**
 * 8.js
 * Linked Lists
 * https://github.com/ashenm/dsa
 *
 * Implements a priority queue abstraction
 *
 * Ashen Gunaratne
 * mail@ashenm.ml
 *
 */

'use strict';

class Node {

  constructor (value, priority, next=null) {
    this.value = value;
    this.priority = priority;
    this.next = next;
  }

}

class LinkedList {

  constructor (head=null) {
    this.head = head;
  }

  // conjoins a new node
  insert (value, priority) {

    let elect;
    let previous;

    const inset = new Node(value, priority);

    if (!this.head) {
      return this.head = inset, this;
    }

    elect = this.head;

    while (elect) {

      if (elect.priority < priority) {
        break;
      }

      previous = elect;
      elect = elect.next;

    }

    inset.next = elect;

    if (!previous) {
      return this.head = inset, this;
    }

    previous.next = inset;

    return this;

  }

  // removes the highest priority node
  serve () {

    const elect = this.head;

    if (!elect) {
      return elect;
    }

    this.head = elect.next;

    return elect;

  }

  // returns the highest priority node
  peak () {
    return this.head;
  }

  // computes the node count
  range () {

    let n = 0;
    let elect = this.head;

    while (elect) {
      elect = elect.next;
      n += 1;
    }

    return n;

  }

}

const main = function () {

  const list = new LinkedList();

  list.insert(25, 20);
  list.insert(50, 30);
  list.insert(75, 25);

  console.log(list.range());
  console.log(JSON.stringify(list.peak()));

  list.serve();

  for (let elect = list.head; elect; elect = elect.next) {
    console.log(elect.value);
  }

};

if (require.main === module) {
  main();
}

// vim: set expandtab shiftwidth=2:
